### DISCUSSION TITLE
Which NuGet.exe do dotnet.exe and MSBuild.exe use?

### DISCUSSION THREAD

**Discussion by Bartleby2718** [NuGet/Home#13537] [Q&A] [answered]
I use .NET 8.0.2 on Visual Studio 2022 on Windows. I have basically the same question but for both `dotnet.exe` and `MSBuild.exe`.

### 1. `dotnet`
`dotnet restore` allows you to restore NuGet packages. I suppose it uses `NuGet.exe` under the hood. If I run `dotnet nuget --version`, I get:
```plaintext
NuGet Command Line
6.9.1.3
```
**Where is this value coming from? Is there a `NuGet.exe` bundled with the .NET SDK?**

FWIW  `where dotnet` gives me `C:\Program Files\dotnet\dotnet.exe`, but I didn't find any `NuGet.exe` in `C:\Program Files\dotnet\sdk\8.0.205`.

### 2. MSBuild
`MSbuild.exe -t:Restore` allows you to restore NuGet packages. I suppose it also uses `NuGet.exe`, but I'm not sure where  it is. I see that `C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\MSBuild.exe` is bundled with VS 2022, but I don't see a `NuGet.exe`. 

Thanks in advance!

**Comment by Bartleby2718** [comment]
@zivkan Do you happen to know the answer to this question as well?

**Comment by zivkan** [comment, marked as answer]
no, the .NET SDK and VS/msbuild does not ship nuget.exe.

NuGet.exe is a ILMerge of a bunch of `NuGet.*.dll`, plus a few additional nuget.exe specific commands. VS/MSBuild and the .NET SDK have all those `NuGet.*.dll`s (easy to find in the .NET SDK install directory), and they use these dlls directly.

**Reply by Bartleby2718** [reply]
Thanks @zivkan!

Sharing more details for future reference:
- For VS 2022: I see 6.9.2 if I hover all any `NuGet.*.dll` file in `C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\NuGet`, and this matches the version I see for `NuGet Package Manager` when I go to `Help` - `About Microsoft Visual Studio`.
- For the .NET 8 SDK: I see 6.9.1 if I hover all any `NuGet.*.dll` file in `C:\Program Files\dotnet\sdk\8.0.205`, and this matches the version I see when I run `dotnet nuget --version`.

**Reply by chriskloe** [reply]
What do I do when I want to call nuget.exe setapikey? Any other way how to specify that? dotnet nuget does not support this option.
Background: I need to do this on a CI/CD server (jenkins) in order to authenticate at the nuget store (gitlab)

**Reply by zivkan** [reply]
dotnet nuget push has an `--api-key` option.  You can also upvote this issue:
* https://github.com/NuGet/Home/issues/12539

**Reply by chriskloe** [reply]
It seemed I need to set an apikey also for dotnet restore.
I managed to test it: did not work though. Maybe there's something weird with my gitlab configuration.

**Reply by zivkan** [reply]
nuget has only ever used "api keys" for push. If you need credentials for restore, that's different. Info of authenticating to private feeds can be found here: https://learn.microsoft.com/nuget/consume-packages/consuming-packages-authenticated-feeds
